Source/Core/VideoBackends/OGL/PostProcessing.cpp

```diff
@@ -12,270 +12,761 @@
+static const u32 FIRST_INPUT_TEXTURE_UNIT = 9;
+static const u32 UNIFORM_BUFFER_BIND_POINT = 4;
+
+static const char* s_vertex_shader = R"(
+out vec2 uv0;
+flat out float layer;
+void main(void)
 {
-	CreateHeader();
+	vec2 rawpos = vec2(gl_VertexID&1, gl_VertexID&2);
+	gl_Position = vec4(rawpos*2.0-1.0, 0.0, 1.0);
+	uv0 = rawpos * src_rect.zw + src_rect.xy;
+	layer = src_layer;
 }
+)";
 
-OpenGLPostProcessing::~OpenGLPostProcessing()
+static const char* s_layered_vertex_shader = R"(
+out vec2 v_uv0;
+void main(void)
 {
-	m_shader.Destroy();
+	vec2 rawpos = vec2(gl_VertexID&1, gl_VertexID&2);
+	gl_Position = vec4(rawpos*2.0-1.0, 0.0, 1.0);
+	v_uv0 = rawpos * src_rect.zw + src_rect.xy;
 }
+)";
+
+static const char* s_geometry_shader = R"(
+
+layout(triangles) in;
+layout(triangle_strip, max_vertices = %d) out;
+
+in vec2 v_uv0[3];
+out vec2 uv0;
+flat out float layer;
 
-void OpenGLPostProcessing::BlitFromTexture(TargetRectangle src, TargetRectangle dst,
-                                           int src_texture, int src_width, int src_height, int layer)
+void main()
 {
-	ApplyShader();
+	for (int i = 0; i < %d; i++)
+	{
+		for (int j = 0; j < 3; j++)
+		{
+			gl_Position = gl_in[j].gl_Position;
+			uv0 = v_uv0[j];
+			layer = float(i);
+			gl_Layer = i;
+			EmitVertex();
+		}
 
-	glBindFramebuffer(GL_DRAW_FRAMEBUFFER, 0);
+		EndPrimitive();
+	}
+}
 
-	glViewport(dst.left, dst.bottom, dst.GetWidth(), dst.GetHeight());
+)";
 
-	OpenGL_BindAttributelessVAO();
+PostProcessingShader::~PostProcessingShader()
+{
+	// Delete texture objects that we own
+	for (RenderPassData& pass : m_passes)
+	{
+		for (InputBinding& input : pass.inputs)
+		{
+			// External textures
+			if (input.texture_id != 0 && input.owned)
+			{
+				glDeleteTextures(1, &input.texture_id);
+				input.texture_id = 0;
+			}
+		}
+
+		if (pass.program != nullptr)
+		{
+			pass.program->Destroy();
+			pass.program.reset();
+		}
+
+		if (pass.output_texture_id != 0)
+		{
+			glDeleteTextures(1, &pass.output_texture_id);
+			pass.output_texture_id = 0;
+		}
+	}
 
-	m_shader.Bind();
+	if (m_framebuffer != 0)
+		glDeleteFramebuffers(1, &m_framebuffer);
+}
 
-	glUniform4f(m_uniform_resolution, (float)src_width, (float)src_height, 1.0f / (float)src_width, 1.0f / (float)src_height);
-	glUniform4f(m_uniform_src_rect, src.left / (float) src_width, src.bottom / (float) src_height,
-		    src.GetWidth() / (float) src_width, src.GetHeight() / (float) src_height);
-	glUniform1ui(m_uniform_time, (GLuint)m_timer.GetTimeElapsed());
-	glUniform1i(m_uniform_layer, layer);
+bool PostProcessingShader::Initialize(const PostProcessingShaderConfiguration* config, int target_layers)
+{
+	m_internal_layers = target_layers;
+	m_config = config;
+	m_ready = false;
 
-	if (m_config.IsDirty())
+	// In case we need to allocate texture objects
+	glActiveTexture(GL_TEXTURE0 + FIRST_INPUT_TEXTURE_UNIT);
+
+	m_passes.reserve(m_config->GetPasses().size());
+	for (const auto& pass_config : m_config->GetPasses())
```

Comment:
Each of these two loops should be their own function; one to do the logic on a single entry, and one that leverages that functions to operate on the sequence.